### PR TITLE
Scheduler

### PR DESCRIPTION
**Pull request by ioquatix**
This is a work in progress.

- [x] Rename `Fiber{}` to `Fiber.schedule{}`.
- [x] Simplify hooks and improve method name consistency. Format is `class_method`, e.g. `io_wait` or `kernel_sleep`.
- [x] Expose scheduler callbacks as functions to aid in correctness, documentation and maintainability.
- [x] Rework `wait.c` to pass `IO` instance, and other adjustments where possible.
- [x] Add support for rescheduling fiber and Mutex support.
- [x] Expose `io_read` as a hook for implementing io_uring (Linux) and IOCP (Windows).
- [x] Add support for ConditionVariable.
- [x] Add support for Queue.
- [x] Rework usage of (some) functions which accept file descriptor arguments.

### CODE REVIEW

**Comment by ioquatix** [comment]
Some notes about `io_uring` after discussion with @dsh0416 who [implemented a scheduler](https://coderemixer.com/2020/08/18/a-walkthrough-of-ruby-3-scheduler/) and has started working on a C interface for `io_uring`.

The most efficient (zero-copy) interface for `IO#read` must go directly to `io_uring`. We schedule the full read operation (buffer, offset, length, etc) and it goes to OS. The OS will inform us when the read operation is completed.

This is different from existing Ruby implementation which uses `IO#wait_readable` in a loop before calling `read`.

However, in practice both interfaces are critical. For example, `libpq` for postgres does not expose some way to hook read, just a way to do non-blocking I/O by checking `IO#wait_...` before calling the `libpq` functions.

So, we need to support both for high performance I/O.

Therefore, I introduced new optional `Scheduler#io_read(io, buffer, offset, length)` and `Scheduler#io_write(io, buffer, offset, length)`. This method is experiment and we will try it with `io_uring` implementation. It should also fit better with Windows IOCP which up until now has been ignored.

File: include/ruby/io.h [hunk 1]
```diff
@@ -113,11 +113,13 @@ typedef struct rb_io_enc_t rb_io_enc_t;
 /* #define FMODE_INET                  0x00400000 */
 /* #define FMODE_INET6                 0x00800000 */
 
-#define GetOpenFile(obj,fp) rb_io_check_closed((fp) = RFILE(rb_io_taint_check(obj))->fptr)
+#define RB_IO_POINTER(obj,fp) rb_io_check_closed((fp) = RFILE(rb_io_taint_check(obj))->fptr)
+#define GetOpenFile RB_IO_POINTER
 
-#define MakeOpenFile(obj, fp) do {\
+#define RB_IO_OPEN(obj, fp) do {\
```
Comment (ioquatix) [inline include/ruby/io.h:127]: @nobu @shyouhei can I get your feedback on this?

File: scheduler.c [hunk 2]
```diff
@@ -59,13 +59,23 @@ VALUE rb_scheduler_io_wait_writable(VALUE scheduler, VALUE io)
...
+int rb_scheduler_supports_io_read(VALUE scheduler)
 {
-    return rb_funcall(scheduler, id_io_read, 4, io, buffer, offset, length);
+    return rb_respond_to(scheduler, id_io_read);
 }
 
-VALUE rb_scheduler_io_write(VALUE scheduler, VALUE io, VALUE buffer, VALUE offset, VALUE length)
+VALUE rb_scheduler_io_read(VALUE scheduler, VALUE io, VALUE buffer, size_t offset, size_t length)
+{
+    return rb_funcall(scheduler, id_io_read, 4, io, buffer, SIZET2NUM(offset), SIZET2NUM(length));
```
Comment (ioquatix) [inline scheduler.c:69]: @shyouhei should we add `RB_SIZET2NUM` etc?

Comment (ioquatix) [reply scheduler.c:69]: Or maybe we can call it `RB_SIZE2NUM` because `SIZET` just looks weird :)

File: include/ruby/io.h [hunk 3]
```diff
@@ -59,6 +59,8 @@ PACKED_STRUCT_UNALIGNED(struct rb_io_buffer_t {
 typedef struct rb_io_buffer_t rb_io_buffer_t;
 
 typedef struct rb_io_t {
+    VALUE self;
```
Comment (ioquatix) [inline include/ruby/io.h:62]: Just for reference, this is a temporary hack.

File: scheduler.c (hunk 2 above)
Comment (dsh0416) [reply scheduler.c:69]: Do we need a `OFFT2NUM`, since `offset` could be a negative number?

**Comment by ioquatix** [comment]
@eregon I have merged your changes into this branch because I want to build on them. Thanks for your effort.

**Comment by ioquatix** [comment]
@jeremyevans do you believe this code is sufficient for your Sequel pool management?

File: scheduler.c (hunk 2 above)
Comment (ioquatix) [reply scheduler.c:83]: Offset is the offset from the start of the buffer, so it should only ever be positive.

File: include/ruby/io.h (hunk 3 above)
Comment (ioquatix) [reply include/ruby/io.h:62]: It turned out to be useful in the `rb_mutex_struct` too, and a common pattern, e.g. `rb_fiber_struct`.

**Comment by jeremyevans** [comment]
I believe Fiber-aware Mutex is a necessary condition for Sequel connection pool management to work correctly with multiple concurrent fibers of the same thread. I'm not sure if ConditionalVariable also needs to be Fiber-aware or not.  So I'm not sure if this will be sufficient by itself, but it is definitely a step in the right direction.

**Comment by ioquatix** [comment]
ConditionalVariable internally uses mutex to synchronize state, so it should be okay, but I'll check it. Thanks for your thoughts.

**Comment by ioquatix** [comment]
I need to check in `thread.c` the following functions:

```
rb_thread_sleep_deadly_allow_spurious_wakeup(void);
rb_thread_wait_for(struct timeval time);
```

and a few others - some are public interface and should go via the same `rb_scheduler_kernel_sleep`.

File: ext/io/console/console.c [hunk 4]
```diff
@@ -513,22 +513,22 @@ console_getch(int argc, VALUE *argv, VALUE io)
...
 	}
 	if (optp->vmin != 1) {
 	    rb_warning("min option ignored");
 	}
 	if (optp->intr) {
-	    w = rb_wait_for_single_fd(fptr->fd, RB_WAITFD_IN, to);
-	    if (w < 0) rb_eof_error();
-	    if (!(w & RB_WAITFD_IN)) return Qnil;
+            VALUE result = RB_NUM2INT(rb_io_wait(io, RUBY_IO_READABLE, timeout));
+            if (result == Qfalse) return Qnil;
```
Comment (ioquatix) [inline ext/io/console/console.c:531]: @nobu what do you think?

**Pull request merged**